File: gateway_server/src/main/kotlin/Balancer.kt
import java.net.InetSocketAddress
import java.nio.channels.SocketChannel

class Balancer {

    var availableServers: ArrayList<SocketChannel> = ArrayList()
    var counter = 0

    fun handle(request: Request): Response {
        when (request) {
            is Request.HiBalancer -> {
                registerServer(request.host, request.port)
                return Response.Pong
            }

            else -> return roundRobin(request)
        }
    }

    private fun registerServer(host: String, port: Int) {
        availableServers.add(
            SocketChannel.open(
                InetSocketAddress(
                    host,
                    port,
                )
            )
        )
    }

    fun roundRobin(request: Request): Response {
        if (request is Request.HiBalancer) {
            registerServer(request.host, request.port)
            return Response.Pong
        }

        val channel = try {
            getNextServer()
        } catch (_: IllegalStateException) {
            return Response.Error("нет вставших серверов(")
        }

        channel.write(request)

        return channel.read() ?: Response.Error("сервер не прислал ответ")
    }

    private fun getNextServer(): GatewayToServersChannel {
        if (availableServers.isEmpty()) {
            throw IllegalStateException()
        }

        val index = counter % availableServers.size

        val candidate = availableServers[index]

        val channel = GatewayToServersChannel(candidate)

        channel.write(Request.Ping)

        if (channel.read() !is Response.Pong) {
            availableServers.removeAt(index)
            return getNextServer()
        }

        counter++

        return channel
    }
}

File: gateway_server/src/balancer.rs
use lab7_shared::{read_frame, write_frame, Request, Response};
use std::net::TcpStream;

#[derive(Clone)]
struct Node {
    host: String,
    port: u16,
}

#[derive(Default)]
pub struct Balancer {
    nodes: Vec<Node>,
    counter: usize,
}

impl Balancer {
    pub fn handle(&mut self, request: Request) -> Response {
        match request {
            Request::HiBalancer { host, port } => {
                self.register_server(host, port);
                Response::Pong
            }
            request => self.proxy(request),
        }
    }

    fn register_server(&mut self, host: String, port: u16) {
        if !self
            .nodes
            .iter()
            .any(|node| node.host == host && node.port == port)
        {
            println!("Registered server {host}:{port}");
            self.nodes.push(Node { host, port });
        }
    }

    fn proxy(&mut self, request: Request) -> Response {
        if self.nodes.is_empty() {
            return Response::Error {
                message: "Нет серверов".to_string(),
            };
        }

        let node = self.nodes[self.counter % self.nodes.len()].clone();
        self.counter += 1;

        match TcpStream::connect((node.host.as_str(), node.port)).and_then(|mut stream| {
            write_frame(&mut stream, &request).map_err(std::io::Error::other)?;
            read_frame::<Response>(&mut stream).map_err(std::io::Error::other)
        }) {
            Ok(response) => response,
            Err(err) => {
                eprintln!("{err:?}");
                Response::Error {
                    message: "Server unavailable".to_string(),
                }
            }
        }
    }
}
